source/base/Promise.ooc

```diff
@@ -8,56 +9,63 @@ _PromiseState: enum {
 }
 
 Promise: abstract class {
-	_state : _PromiseState
+	_state: _PromiseState
 	init: func
 	wait: abstract func -> Bool
+	wait: abstract func ~timeout (seconds: Double) -> Bool
 	cancel: virtual func -> Bool { false }
 	start: static func (action: Func) -> This {
 		_ThreadPromise new(action)
 	}
+	operator + (other: This) -> PromiseCollector {
+		collector := PromiseCollector new()
+		collector add(this)
+		collector add(other)
+		collector
+	}
 }
 
 _ThreadPromise: class extends Promise {
 	_action: Func
 	_thread: Thread
-	_mutex := Mutex new()
+	_threadAlive := true
 	init: func (task: Func) {
 		super()
 		this _state = _PromiseState Unfinished
 		this _action = func {
 			task()
-			this _mutex lock()
 			if (this _state != _PromiseState Cancelled)
 				this _state = _PromiseState Finished
-			this _mutex unlock()
 		}
 		this _thread = Thread new(this _action)
 		this _thread start()
 	}
 	free: override func {
-		if (this _state == _PromiseState Unfinished)
+		//TODO: If we don't want to wait, set a freeOnCompletion flag instead, and check it once the thread task finishes
+		if (this _threadAlive)
 			this _thread wait()
 		this _thread free()
 		(this _action as Closure) dispose()
-		this _mutex free()
 		super()
 	}
 	wait: func -> Bool {
-		if (this _state == _PromiseState Unfinished)
-			this _thread wait()
+		if (this _threadAlive)
+			if (this _thread wait())
```

Comment:
The promise has not been fulfilled, for example if someone cancelled it before it completed.